extensions/amp-selector/0.1/test/test-amp-selector.js

```diff
@@ -1164,6 +1164,38 @@ describes.realWin(
           }
         );
 
+        it(
+          'should update focus when the user presses the home/end keys when ' +
+            'keyboard-select-mode is enabled',
+          () => {
+            const ampSelector = getSelector({
+              attributes: {
+                'keyboard-select-mode': 'focus',
+              },
+              config: {
+                count: 3,
+              },
+            });
+            ampSelector.children[1].setAttribute('selected', '');
+            ampSelector.build();
+            expect(ampSelector.children[0].tabIndex).to.equal(-1);
+            expect(ampSelector.children[1].tabIndex).to.equal(0);
+            expect(ampSelector.children[2].tabIndex).to.equal(-1);
+            return keyPress(ampSelector, Keys.HOME)
```

Comment:
Make this into two separate tests, since you are testing two pieces of functionality. Each one could fail separately, so you should be able to tell which one broke independently.

One problem with combining multiple tests in general is that if you break multiple pieces, you get just one test failure, without quickly knowing what are all the things broken that need to be fixed. This could lead you to make code changes to fix "Home" but not "End". Then you run the tests again only to find "End" does not work. In more complex cases, this could cause changes you made for "Home" to not really be valid and require rework.

Tests also serve as a way of communicating the functionality supported by a component. By having separate test cases, it is more clear that there are two different pieces of functionality here.

Separating test cases also makes it easier to test different conditions. For example, I think it is important to test the "disabled" item skipping here. Doing this in one test case for both home/end is a bit messy. Separating out the tests is better here, for testing the end case, you would

1. Select children[0]
2. Set children[2] as disabled
3. press the end key

and for the other end

1. Select children[2]
2. Set children[0] as disabled
3. press the home key

Trying to do that within this test would require you change the disabled state within the test, which makes it much harder to reason about.

Reply:
That makes sense. Thanks for typing this out!